sql: support table storage params change in the declarative schema changer

shghasemi opened the pull request:
###  sql: support table storage params change in the declarative schema changer

Previously, storage parameters were set and reset using the legacy schema
changer. With this change, setting most non-backfilling table storage params
will use the declarative schema changer. Setting ttl-related params, schema_locked,
and infer_rbr_region_col_using_constraint is not implemented yet.

### schemachanger: handle RESET storage param in declarative schema changer

This reuses the logic that was added to handle SET. The only difference
is that RESET always drops the StorageParam element.

###  schemachanger: handle manual schema_locked setting in declarative

This storage parameter is special, so this requires logic to use the
pre-existing TableSchemaLocked element rather than the normal storage
param element.

As part of this, we partially revert bc3f228 and refactor the check to be a
little simpler since we don't actually need to handle the mixed version
case now.

Epic CRDB-31281
Fixes cockroachdb#155990

Release note: None

shghasemi committed:
sql: refactor storage param set and reset function

Epic CRDB-31281

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/158390)
<!-- Reviewable:end -->

shghasemi committed:
sql: support table storage params change in the declarative schema changer

Previously, storage parameters were set and reset using the legacy schema
changer. With this change, setting most non-backfilling table storage params
will use the declarative schema changer. Setting ttl-related params, schema_locked,
and infer_rbr_region_col_using_constraint is not implemented yet.

Epic CRDB-31281

Fixes #155990

Release note: None

fqazi reviewed (changes requested):
@shghasemi Great work!

Just some very minor comments on this.

> @fqazi reviewed 1 of 1 files at r1, 31 of 31 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-:-Of_pFhu1YYhpeoR43jd:brtqvjl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shghasemi)

___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_nmjbE1oeH-gQfh-r:-Of_nmjbE1oeH-gQfh-s:b-454mq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/scdecomp/decomp.go#L507)):*
> ```Go
> 			continue
> 		}
> 		if key == "schema_locked" {
> ```

Lets have a comment explaining schema_locked is handled seperately.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_set_storage_param.go` line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_n68v-hOVpHKOb0gj:-Of_n68v-hOVpHKOb0gk:bt35exo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_set_storage_param.go#L95)):*
> ```Go
> // where tbl is the referenced table.
> func isTableReferencedByFK(b BuildCtx, tbl *scpb.Table) bool {
> 	count := 0
> ```

Nit: Lets just flip a boolean
___
*[`toString.diff` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_lX2K-nDdlXuvP6PB:-Of_lX2K-nDdlXuvP6PC:b3noawf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/toString.diff#L3)):*
> ```Diff
> diff --git a/pkg/sql/storageparam/tablestorageparam/table_storage_param.go b/pkg/sql/storageparam/tablestorageparam/table_storage_param.go
> index c4dcd99c39a..edf6e15bb9e 100644
> --- a/pkg/sql/storageparam/tablestorageparam/table_storage_param.go
> ```

Accidentally added file?
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_p7nd4xAfQFFS-ydD:-Of_p7ne7l0WUCcopY5A:b-bm50pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain#L6)):*
> ```
> 
> /* test */
> EXPLAIN (DDL) ALTER TABLE t SET (exclude_data_from_backup=true);
> ```

Can we set multiple at the same time?
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_oqOF5NYqj3yzCbyR:-Of_oqOF5NYqj3yzCbyS:byt3fh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain#L14)):*
> ```
>  │         │    └── ABSENT → PUBLIC TableStorageParam:{DescID: 104 (t), Name: "exclude_data_from_backup", Value: "true"}
>  │         ├── 1 element transitioning toward TRANSIENT_PUBLIC
>  │         │    └── PUBLIC → ABSENT TableSchemaLocked:{DescID: 104 (t)}
> ```

We shouldn't need a table schema locked toggle for storage parameters. Unless we are adding columns or some type of backfilling op.


<!-- Sent from Reviewable.io -->

rafiss committed:
schemachanger: handle RESET storage param in declarative schema changer

This reuses the logic that was added to handle SET. The only difference
is that RESET always drops the StorageParam element.

Release note: None

rafiss committed:
schemachanger: handle manual schema_locked setting in declarative

This storage parameter is special, so this requires logic to use the
pre-existing TableSchemaLocked element rather than the normal storage
param element.

As part of this, we partially revert
bc3f228e69fb21e88be02dfb75e8c088c7a1e373 and refactor the check to be a
little simpler since we don't actually need to handle the mixed version
case now.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158390)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @mgartner)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_set_storage_param.go` line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_n68v-hOVpHKOb0gj:-Og30uwj2X8DG8-wh9pv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_set_storage_param.go#L95)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Lets just flip a boolean
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_nmjbE1oeH-gQfh-r:-Og313pY5IwOfA7iQk54:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/scdecomp/decomp.go#L507)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Lets have a comment explaining schema_locked is handled seperately.
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_p7nd4xAfQFFS-ydD:-Og315747VRALY8dWzDq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain#L6)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Can we set multiple at the same time?
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_oqOF5NYqj3yzCbyR:-Og2zaCD7y9FsKJRBjQS:b-gvcnc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/pkg/sql/schemachanger/testdata/end_to_end/alter_table_set_storage_param/alter_table_set_storage_param.explain#L14)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We shouldn't need a table schema locked toggle for storage parameters. Unless we are adding columns or some type of backfilling op.
</blockquote></details>

that's true, but i think i'm more in favor of keeping the schema_locked toggle on for all ALTER TABLE operations no matter what, in order to keep the code simpler. the same reasoning could be used for ALTER TABLE .. RENAME COLUMN, and currently we also toggle schema_locked for that too.

if we determine later that we need to optimize these metadata only operations to avoid toggling schema_locked, we can make the change later
___
*[`toString.diff` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-Of_lX2K-nDdlXuvP6PB:-Og318IFDe0zKJAWqSNw:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1c0119d705c894500c56a0d0deba1958f367805/toString.diff#L3)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Accidentally added file?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
This all lookgs great @rafiss !

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> @fqazi reviewed 30 of 30 files at r5, 32 of 32 files at r6, 7 of 7 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158390#-:-Og44qB68pHWyEGfCiS4:bx2j4j9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

rafiss commented:
TFTR!

bors r+
